include/rapidjson/document.h

```diff
@@ -2172,6 +2333,145 @@ GenericValue<Encoding,Allocator>::GenericValue(const GenericValue<Encoding,Sourc
     }
 }
 
+//! Helper class for accessing Value of array type.
+/*!
+    Instance of this helper class is obtained by \c GenericValue::GetArray().
+    In addition to all APIs for array type, it provides range-based for loop if \c RAPIDJSON_HAS_CXX11_RANGE_FOR=1.
+*/
+template <bool Const, typename ValueT>
+class GenericArray {
+public:
+    typedef GenericArray<true, ValueT> ConstArray;
+    typedef GenericArray<false, ValueT> Array;
+    typedef ValueT PlainType;
+    typedef typename internal::MaybeAddConst<Const,PlainType>::Type ValueType;
+    typedef ValueType* ValueIterator;  // This may be const or non-const iterator
+    typedef const ValueT* ConstValueIterator;
+    typedef typename ValueType::AllocatorType AllocatorType;
+    typedef typename ValueType::StringRefType StringRefType;
+
+    template <typename, typename>
+    friend class GenericValue;
+
+    GenericArray() : ptr_() {}
+    GenericArray(const GenericArray& rhs) : ptr_(rhs.ptr_) {}
+    GenericArray& operator=(const GenericArray& rhs) { ptr_ = rhs.ptr_; return *this; }
+    ~GenericArray() {}
+
+    SizeType Size() const { return ptr_->Size(); }
+    SizeType Capacity() const { return ptr_->Capacity(); }
+    bool Empty() const { return ptr_->Empty(); }
+    void Clear() const { ptr_->Clear(); }
+    ValueType& operator[](SizeType index) const {  return (*ptr_)[index]; }
+    ValueIterator Begin() const { return ptr_->Begin(); }
+    ValueIterator End() const { return ptr_->End(); }
+    const GenericArray& Reserve(SizeType newCapacity, AllocatorType &allocator) const { ptr_->Reserve(newCapacity, allocator); return *this; }

```

Comment:
As the wrappers are only a plain pointer anyway (i.e. coying is as cheap as it gets), we can return by value here.  This can help compilers with optimization as it reduced one indirection.

But, more importantly: Why is `Reserve` declared `const`?